### ISSUE TITLE
Trainer, fails to load.

### ISSUE THREAD

**Issue by MarcusLoppe** [lucidrains/meshgpt-pytorch#28]
When loading a model using the current version, the optimizer fails to load since it excepts a dict with the optimizer key. 

So rename:
        self.optimizer.load_state_dict(pkg['optimizer'])
To:
        self.optimizer.load_state_dict(pkg)

```
    def load(self, path):
        path = Path(path)
        assert path.exists()

        pkg = torch.load(str(path))

        if version.parse(__version__) != version.parse(pkg['version']):
            self.print(f'loading saved mesh transformer at version {pkg["version"]}, but current package version is {__version__}')

        self.model.load_state_dict(pkg['model'])
        self.optimizer.load_state_dict(pkg['optimizer'])
        self.step.copy_(pkg['step'])
```

**Comment by lucidrains** [comment]
@MarcusLoppe oh are you loading an old checkpoint?

**Comment by lucidrains** [comment]
@MarcusLoppe i changed the way i handled optimizer, schedulers and warmup [here](https://github.com/lucidrains/pytorch-custom-utils/blob/main/pytorch_custom_utils/optimizer_scheduler_warmup.py), so older checkpoints will not be compatible unfortunately

**Comment by MarcusLoppe** [comment]
No, I'm using the latest.

The issue is that the optimizer  load_state_dict accepts **pkg** dict but in trainer you pass only the optimizer key of pkg, not the entire pkg dict  **pkg['optimizer']** 
  
optimizer_scheduler_warmup.py:
```
  def load_state_dict(self, pkg):
        self.optimizer.load_state_dict(pkg['optimizer'])
        self.scheduler.load_state_dict(pkg['scheduler'])
        self.warmup.load_state_dict(pkg['warmup'])
```
        
 trainer.py
  ```
  def load(self, path):
        path = Path(path)
        assert path.exists()

        pkg = torch.load(str(path))

        if version.parse(__version__) != version.parse(pkg['version']):
            self.print(f'loading saved mesh autoencoder at version {pkg["version"]}, but current package version is {__version__}')

        self.model.load_state_dict(pkg['model'])
        self.ema_model.load_state_dict(pkg['ema_model'])
        self.optimizer.load_state_dict(pkg['optimizer'])
```

**Comment by lucidrains** [comment]
@MarcusLoppe i also changed the way it is saved [here](https://github.com/lucidrains/pytorch-custom-utils/blob/main/pytorch_custom_utils/optimizer_scheduler_warmup.py#L48)

**Comment by lucidrains** [comment]
this runs for me

```python
import torch

from meshgpt_pytorch import (
    MeshAutoencoder,
    MeshTransformer,
    MeshAutoencoderTrainer,
    MeshTransformerTrainer,
    DatasetFromTransforms
)

# autoencoder

autoencoder = MeshAutoencoder(
    num_discrete_coors = 128,
    local_attn_encoder_depth = 0,
    local_attn_decoder_depth = 2
)

# mock dataset

from torch.utils.data import Dataset

class MockDataset(Dataset):
    def __init__(self):
        pass

    def __len__(self):
        return 100

    def __getitem__(self, idx):
        from random import randrange
        return torch.randn(randrange(20, 40), 3), torch.randint(0, 10, (randrange(20, 40), 3))

dataset = MockDataset()

trainer = MeshAutoencoderTrainer(
    autoencoder,
    dataset = dataset,
    val_dataset = dataset,
    batch_size = 2,
    grad_accum_every = 2,
    val_every = 1,
    num_train_steps = 1,
    checkpoint_every = 5,
    accelerator_kwargs = dict(
        cpu = True
    )
)

trainer.save('./autoencoder.pt')
trainer.load('./autoencoder.pt')
trainer()
```

**Comment by MarcusLoppe** [comment]
> this runs for me
> 
> ```python
> import torch
> 
> from meshgpt_pytorch import (
>     MeshAutoencoder,
>     MeshTransformer,
>     MeshAutoencoderTrainer,
>     MeshTransformerTrainer,
>     DatasetFromTransforms
> )
> 
> # autoencoder
> 
> autoencoder = MeshAutoencoder(
>     num_discrete_coors = 128,
>     local_attn_encoder_depth = 0,
>     local_attn_decoder_depth = 2
> )
> 
> # mock dataset
> 
> from torch.utils.data import Dataset
> 
> class MockDataset(Dataset):
>     def __init__(self):
>         pass
> 
>     def __len__(self):
>         return 100
> 
>     def __getitem__(self, idx):
>         from random import randrange
>         return torch.randn(randrange(20, 40), 3), torch.randint(0, 10, (randrange(20, 40), 3))
> 
> dataset = MockDataset()
> 
> trainer = MeshAutoencoderTrainer(
>     autoencoder,
>     dataset = dataset,
>     val_dataset = dataset,
>     batch_size = 2,
>     grad_accum_every = 2,
>     val_every = 1,
>     num_train_steps = 1,
>     checkpoint_every = 5,
>     accelerator_kwargs = dict(
>         cpu = True
>     )
> )
> 
> trainer.save('./autoencoder.pt')
> trainer.load('./autoencoder.pt')
> trainer()
> ```

Hmm, I loaded model I trained over night, might due to the update in the pytorch-custom-utils.
But alright If you tested it then I'll close it. :)

**Comment by lucidrains** [comment]
yea my bad, this may the last breaking change

the repository is almost finished
